feat!: skip running warnings in --quiet mode

me4502 opened the pull request:
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

https://github.com/eslint/rfcs/tree/main/designs/2023-only-run-reporting-rules

Fixes #16450

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

#### What changes did you make? (Give an overview)

This PR implements the RFC around only running rules that actually report information in quiet mode. This adds a predicate function that is passed down to the runRules function that filters out all non-error rules when the `--quiet` flag is entered and `maxWarnings` is not in use. All rules returning false from the predicate function are skipped during rule running.

TODO:
- [x] Tests
- [x] Documentation changes
- [X] Ignoring of unused disable directives

I believe the `--max-warnings` flag handling is already covered by the existing tests, as from a test perspective it only cares about if it still works with quiet mode (everything else is an implementation detail). I've added tests that the `ruleFilter` option does actually filter out rules to the linter.

#### Is there anything you'd like reviewers to focus on?

Rec0iL99 commented:
Not stale. Work in progress.

me4502 commented:
Sorry am still working on it; have been a bit busy and given this is blocked by the next major release it’s not too high on my list of priorities as I’m assuming that’s still at least a little while away - I’ll try to finish it up sometime soonish then rebase it every so often to keep the bot happy

me4502 commented:
Not stale sorry, I’ve been fairly busy

me4502 commented:
I believe this should be ready for review. I looked over it and wrote some tests to validate that the `ruleFilter` option on the linter functions correctly, and the interactions with the `--max-warnings` flag appear to be fully covered by existing tests

nzakas commented:
@me4502 we are starting to look at v9 work. It looks like you're still updating this PR. Do you think it's in a good state to review?

nzakas commented:
Ah got it, thanks for keeping up!

mdjermanovic commented:
Converting to draft just so we don't accidentally merge this before we start with v9 prereleases.

nzakas commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -159,6 +159,8 @@ The `ESLint` constructor takes an `options` object. If you omit the `options` ob
   Default is `[]`. An array of paths to directories to load custom rules from.
 * `options.useEslintrc` (`boolean`)<br>
   Default is `true`. If `false` is present, ESLint doesn't load configuration files (`.eslintrc.*` files). Only the configuration of the constructor options is valid.
+* `options.ruleFilter` (`Function`)<br>
```
For clarity, maybe this is better?
```suggestion
* `options.ruleFilter` (`({ruleId: string, severity: number}) => boolean`)<br>
```

nzakas commented on docs/src/use/command-line-interface.md:
```diff
@@ -466,6 +466,8 @@ This option allows you to specify a warning threshold, which can be used to forc
 
 Normally, if ESLint runs and finds no errors (only warnings), it exits with a success exit status. However, if `--max-warnings` is specified and the total warning count is greater than the specified threshold, ESLint exits with an error status.
 
+When used alongside `--quiet`, this will cause rules marked as warn to still be run, but not reported.
```
Let's emphasize this point.

```suggestion
::: important
When used alongside `--quiet`, this will cause rules marked as warn to still be run, but not reported.
:::
```

nzakas commented on lib/cli.js:
```diff
@@ -178,7 +189,8 @@ async function translateOptions({
         overrideConfigFile,
-        reportUnusedDisableDirectives: reportUnusedDisableDirectives ? "error" : void 0
+        reportUnusedDisableDirectives: reportUnusedDisableDirectives ? "error" : void 0,
+        ruleFilter: quiet && maxWarnings === 0 ? quietRuleFilter : () => true
```
Can we add a comment here just to clarify why this is happening?

nzakas commented on lib/linter/apply-disable-directives.js:
```diff
@@ -456,10 +463,31 @@ module.exports = ({ directives, disableFixes, problems, reportUnusedDisableDirec
-    return reportUnusedDisableDirectives !== "off"
-        ? lineDirectivesResult.problems
-            .concat(blockDirectivesResult.unusedDirectives)
+    if (reportUnusedDisableDirectives !== "off") {
```
Can we add some comment in this block explaining each step?

nzakas commented on lib/linter/apply-disable-directives.js:
```diff
@@ -456,10 +463,31 @@ module.exports = ({ directives, disableFixes, problems, reportUnusedDisableDirec
+            .filter(unusedDirective => {
+                const ruleId = unusedDirective.ruleId;
+
+                if (ruleId === null && rulesToIgnore.size > 0 || rulesToIgnore.has(ruleId)) {
```
It seems like we could simplify this by adding a `null` to `rulesToIgnore` and then just checking `rulesToIgnore.has(ruleId)`?

Also, a comment here explaining why we are filtering out rules without a `ruleId` would be helpful.

nzakas commented on tests/lib/linter/linter.js:
```diff
@@ -12391,6 +12391,53 @@ describe("Linter with FlatConfigArray", () => {
+                        "alert(\"test\");"
+                    ].join("\n");
+                    const config = {
+                        rules: { "no-alert": 1 }
```
Can you add a test that has multiple rules and only one or two filtered out? We want to be sure that the filter is working properly.

nzakas reviewed (changes requested):
Thanks for putting this together! Overall, this looks good. I left a few notes where it would be helpful to add some comments.

Also, we will need tests for the changes in `cli.js`, `eslint.js`, and `flat-eslint.js` to ensure we're properly testing this change.

me4502 commented:
I've just addressed some of the feedback and will complete the test-related feedback shortly, thanks.

I just have one question on the `cli.js` test. I initially did not write one as there's already a test for max warnings still functioning under `quiet` mode, which was the main piece of functionality I felt made sense to test in this file. Is this one sufficient? If not, is there anything in particular that you're wanting a test for here? I'm not super familiar with chai/sinon/proxyquire as a test framework, but it currently looks like all tests in that file are only really asserting on exit codes and logs which wouldn't be impacted by these rules not running.

nzakas commented:
I believe you're referring to this test?

https://github.com/eslint/eslint/blob/de165c108203c6703516ac651f5b4cac5b241804/tests/lib/cli.js#L707-L714

If so, I think you're correct in that this test will properly cover these changes and we don't need any additional tests.

me4502 commented:
Yeah that's the one I was referring to. I've just added a few extra tests as requested, and also made the error message follow the same format as the others (I missed a `.` at the end)

mdjermanovic commented:
> but it currently looks like all tests in that file are only really asserting on exit codes and logs which wouldn't be impacted by these rules not running

We could make a test with a rule that always throws an error in `create()`. Exit code `0` would be good enough proof that the rule wasn't run.

mdjermanovic commented on lib/eslint/eslint.js:
```diff
@@ -300,7 +305,8 @@ function processOptions({
         rulePaths,
-        useEslintrc
+        useEslintrc,
+        ruleFilter
```
Looks like the `ESLint` class accepts the new option, but then doesn't use it? We should either implement the change in this class as well, or not allow passing the new option. Per @nzakas's https://github.com/eslint/eslint/issues/16450#issuecomment-1310829909, I believe the idea was to not implement this change in the `ESLint` class, so it should not allow this option to be passed to the constructor.

me4502 commented:
I've just applied the code changes, thanks.

As for the test, is there currently something setup in the testing framework to allow a test rule as you've described? From what I can tell from the cli tests, the only way to add a rule that is accessible is either a config file or a `--rule` flag, rather than directly adding code for a rule.

mdjermanovic commented:
> As for the test, is there currently something setup in the testing framework to allow a test rule as you've described? From what I can tell from the cli tests, the only way to add a rule that is accessible is either a config file or a `--rule` flag, rather than directly adding code for a rule.

Since this change affects only flat config mode, we could add a flat config file with a rule directly in it.

For example, `tests/fixtures/eslint.config-rule-throws.js`:

```js
module.exports = [
    {
        plugins: {
            foo: {
                rules: {
                    bar: {
                        create() {
                            throw new Error("Boom!");
                        }
                    }
                }
            }
        }
    },
    {
        rules: {
            "foo/bar": "warn"
        }
    }
];

```

And then add a test that uses this config, for example in [tests](https://github.com/eslint/eslint/blob/48ed5a6dad478a14d3e823f137455c523f373e0b/tests/lib/cli.js#L569-L593) for the `--quiet` mode:

```js
if (useFlatConfig) {
    it(`should not run rules set to 'warn' with configType:${configType}`, async () => {
        const filePath = getFixturePath("single-quoted.js");
        const configPath = getFixturePath("eslint.config-rule-throws.js");
        const cliArgs = `--quiet --config ${configPath}' ${filePath}`;

        const exit = await cli.execute(cliArgs, null, useFlatConfig);

        assert.strictEqual(exit, 0);
    });
}
```

mdjermanovic commented on lib/cli.js:
```diff
@@ -178,7 +189,13 @@ async function translateOptions({
+         * For performance reasons rules not marked as 'error' are filtered out in quiet mode. As maxWarnings
+         * requires rules set to 'warn' to be run, we only filter out 'warn' rules if maxWarnings is 0.
+         */
+        ruleFilter: quiet && maxWarnings === -1 ? quietRuleFilter : () => true
```
Since only the `FlatESLint` class accepts this option, we should set it below, inside `if (configType === "flat")`.

mdjermanovic commented:
One thing we forgot to mention in the RFC is caching.

If user runs `eslint --cache --quiet`, only errors will be stored in the cache because rules set to warn were not run. If the user then runs  just `eslint --cache` without `--quiet`, results pulled from the cache can be wrong for this command because they don't include warnings that the code might have.

I don't have a concrete solution at the moment, just wanted to note that we likely need to address this.

fasttime commented:
> If user runs `eslint --cache --quiet`, only errors will be stored in the cache because rules set to warn were not run. If the user then runs just `eslint --cache` without `--quiet`, results pulled from the cache can be wrong for this command because they don't include warnings that the code might have.
> 
> I don't have a concrete solution at the moment, just wanted to note that we likely need to address this.

There is this function `hashOfConfigFor` which is used to calculate a hash for the config of each file, so if the config changes, the hash changes, too, and the cache results are discarded.

https://github.com/eslint/eslint/blob/98926e6e7323e5dd12a9f016cb558144296665af/lib/cli-engine/lint-result-cache.js#L116-L129

Maybe we could extend that function to also include the `quiet` setting in the hash.

mdjermanovic commented:
> Maybe we could extend that function to also include the `quiet` setting in the hash.

A problem with this solution is that the `quiet` setting is a function (`ruleFilter`), so it's not serializable. We could `.toString()` it, but that wouldn't be fully reliable.

fasttime commented:
> > Maybe we could extend that function to also include the `quiet` setting in the hash.
> 
> A problem with this solution is that the `quiet` setting is a function (`ruleFilter`), so it's not serializable. We could `.toString()` it, but that wouldn't be fully reliable.

True. Even if the rule filter function stays the same, it could still report different values depending on user-defined runtime settings. So probably it's just not reliable to filter the results before writing them to the cache when `ruleFilter` is a custom function.

me4502 commented:
Could an option be to save a list of run rules to the cache, and if it doesn't match act as if there's no valid cache? Does mean having to iterate over the config and run the function for the given rules though

nzakas commented:
> One thing we forgot to mention in the RFC is caching.
> 
> If user runs `eslint --cache --quiet`, only errors will be stored in the cache because rules set to warn were not run. If the user then runs just `eslint --cache` without `--quiet`, results pulled from the cache can be wrong for this command because they don't include warnings that the code might have.
> 
> I don't have a concrete solution at the moment, just wanted to note that we likely need to address this.

I think this behavior is acceptable. If you're running ESLint with different command line flags, it seems like different results should be expected? But I'm guessing most people don't change the flags they're using very often.

me4502 commented:
I've just made the requested change to the cli file, as well as added the requested test alongside another one that validates that the config with the rule that throws on creation actually gets created when `--max-warnings` is also present

mdjermanovic commented on lib/cli.js:
```diff
@@ -193,6 +204,12 @@ async function translateOptions({
+
+        /*
+         * For performance reasons rules not marked as 'error' are filtered out in quiet mode. As maxWarnings
+         * requires rules set to 'warn' to be run, we only filter out 'warn' rules if maxWarnings is 0.
```
```suggestion
         * requires rules set to 'warn' to be run, we only filter out 'warn' rules if maxWarnings is not specified.
```

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -12502,6 +12502,69 @@ describe("Linter with FlatConfigArray", () => {
+
+                    assert.strictEqual(messages.length, 1);
+                });
+            });
```
Can you also add some tests with eslint disable directives?

me4502 replied:
I've just added some tests here that should cover all permutations. I had to refactor a bit of the added disable directive handling as it seems there were some other refactors done that made my code no longer work where it was originally positioned.

nzakas commented on lib/linter/linter.js:
```diff
@@ -1391,7 +1405,8 @@ class Linter {
                 .sort((problemA, problemB) => problemA.line - problemB.line || problemA.column - problemB.column),
-            reportUnusedDisableDirectives: options.reportUnusedDisableDirectives
+            reportUnusedDisableDirectives: options.reportUnusedDisableDirectives,
+            configuredRules
```
I agree, I don't think we want this running in `Linter` in eslintrc mode. Let's keep the changes in the flat config logic.

me4502 commented:
As it seems some breaking changes for v9 are now in the main branch, should I mark this ready for review?

mdjermanovic commented:
> As it seems some breaking changes for v9 are now in the main branch, should I mark this ready for review?

Yes, I marked this as ready for review now.

mdjermanovic commented on lib/linter/apply-disable-directives.js:
```diff
@@ -443,17 +450,33 @@ module.exports = ({ directives, disableFixes, problems, reportUnusedDisableDirec
+    // This determines a list of rules that are not being run by the given ruleFilter, if present.
+    const rulesToIgnore = configuredRules && ruleFilter
+        ? new Set(Object.keys(configuredRules).filter(ruleId => {
+            const severity = ConfigOps.getRuleSeverity(configuredRules[ruleId]);
+
+            return !ruleFilter({ severity, ruleId });
```
```suggestion
            const severity = ConfigOps.getRuleSeverity(configuredRules[ruleId]);
            
            if (severity === 0) {
                return false;
            }

            return !ruleFilter({ severity, ruleId });
```

Rules that are turned off should not be ignored when checking unused disable directives (and generally shouldn't be passed to `ruleFilter()`).

```js
/* eslint no-undefined: 0 */

/* eslint-disable no-undefined */   // <- should be reported as unused 

```

mdjermanovic commented:
> > One thing we forgot to mention in the RFC is caching.
> > If user runs `eslint --cache --quiet`, only errors will be stored in the cache because rules set to warn were not run. If the user then runs just `eslint --cache` without `--quiet`, results pulled from the cache can be wrong for this command because they don't include warnings that the code might have.
> > I don't have a concrete solution at the moment, just wanted to note that we likely need to address this.
> 
> I think this behavior is acceptable. If you're running ESLint with different command line flags, it seems like different results should be expected? But I'm guessing most people don't change the flags they're using very often.

Okay, I think it's acceptable but it would be better to avoid it if possible. Something like @me4502 suggested in https://github.com/eslint/eslint/pull/17274#issuecomment-1840790604 might be the solution. However, given the complexity, I think this should be revisited at some later point (not in this PR). Also, it seems that the same problem exists with some other CLI options (e.g., `--no-inline-config`) so we could try to fix that altogether.

mdjermanovic reviewed (approved):
LGTM, thanks! Leaving open for @nzakas to verify.

nzakas commented on tests/fixtures/eslint.config_rule_throws.js:
Can we rename to `eslint.config-rule-throws.js`? We generally don't have underscores in filenames.

nzakas reviewed (approved):
Because I'm blocked, I'm going to go ahead and merge this and then handle the file renaming myself.

Thanks so much for sticking with this feature @me4502! 🙏

me4502 commented:
Thanks :) Sorry I named the file with underscores as the ones around it had those too
![image](https://github.com/eslint/eslint/assets/546754/113aa99e-bcf9-4157-bb0d-2799d55ca711)

nzakas commented:
Yeah no worries, I think those snuck in at some point so I understand the confusion.
